<a href="https://colab.research.google.com/github/suzannelittle/dcu-dmv/blob/main/Intro_to_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction to Pandas
*Suzanne Little, suzanne.little@dcu.ie*

1. What does pandas offer? `DataFrame` & `Series`
2. Reading in data (I/O)
3. View your data
4. Understand your data
5. Manipulate the dataframe

## `DataFrames` & `Series`

## Resources

* "10mins to Pandas" - https://pandas.pydata.org/docs/user_guide/10min.html
* Jupyter notebook with exercises - https://github.com/dlab-berkeley/introduction-to-pandas
* DataCamp - https://www.datacamp.com/courses/pandas-foundations
* Pandas cheatsheets
 - https://www.datacamp.com/community/blog/python-pandas-cheat-sheet
 - https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf


In [ ]:
import pandas as pd
%matplotlib inline

Convention is to use `pd` as the shorthand for pandas. Similarly `np` is the shorthand for numpy which is another key data science python library.

[Pandas](https://pandas.pydata.org) provides a way to store, manipulate ("wrangle"), clean and analyse your data. It is based around an indexed table structure - "DataFrame" - that is made up of indexed lists - "Series".

![Diagram of Series and DataFrame](https://storage.googleapis.com/lds-media/images/series-and-dataframe.width-1200.png "Series plus Series equals DataFrame")

Diagram above is from https://www.learndatasci.com/tutorials/python-pandas-tutorial-complete-introduction-for-beginners/

In [ ]:
data = {
    'apples': [3, 2, 0, 1],
    'oranges': [0, 3, 7, 2]
}

In [ ]:
purchases = pd.DataFrame(data)

In [ ]:
purchases

apples  oranges
0       3        0
1       2        3
2       0        7
3       1        2

In [ ]:
type(purchases)

pandas.core.frame.DataFrame

In [ ]:
purchases = pd.DataFrame(data, index=['June', 'Robert', 'Lily', 'David'])

In [ ]:
purchases

apples  oranges
June         3        0
Robert       2        3
Lily         0        7
David        1        2

In [ ]:
june = purchases.loc['June']
print(type(june))
print(june)

<class 'pandas.core.series.Series'>
apples     3
oranges    0
Name: June, dtype: int64


In [ ]:
se = purchases["apples"]
print(type(se))
print(se)

<class 'pandas.core.series.Series'>
June      3
Robert    2
Lily      0
David     1
Name: apples, dtype: int64


## Reading data in (I/O)

## read_csv    
https://pandas.pydata.org/pandas-docs/stable/io.html#io-read-csv-table

Read the DataSampler.csv file into a pandas dataframe object. Convention is to call dataframes `df`.    
The main argument to `read_csv` is the path to the file. If you are running this notebook using Jupyter on your computer then this can be a local path (ie, you've downloaded DataSampler.csv) or a url.

In [ ]:
df_csv = pd.read_csv("https://raw.githubusercontent.com/suzannelittle/ca682i/master/data/sampler/DataSampler.csv")  # look at the other arguments to read_csv. Anything you could add?

In [ ]:
df_csv.head()  # show the first 5 lines

A         B   C   D       E     F           G          H  \
0  03/10/2016 16:00:00  0.662052  16  58  banana  AYPI  Rossan Pt.  54.700000   
1  03/10/2016 16:01:00  0.123787  43  57   apple  RGVS  Mullet Pen  54.216667   
2  03/10/2016 16:02:00  0.032207  62  66  banana  JOBI        Erne  54.500000   
3  03/10/2016 16:03:00  0.865759  78  60  celery  MBPL     Kilrush  52.633333   
4  03/10/2016 16:04:00  0.321649  78  52  banana  MFRO       Louth  53.083333   

           I  
0  -8.783333  
1 -10.033333  
2  -8.266667  
3  -9.483333  
4  -6.533333

In [ ]:
df_csv.shape  # find out how many rows and columns

(151, 9)

So the DataSampler has 151 entries (rows) with 9 attributes (columns) each. What are the attributes?

In [ ]:
df_csv.columns

Index(['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'], dtype='object')

So using .columns will get you the list of attribute names (the column headers). You can also change the column names by changing this.

In [ ]:
df_csv.columns = ["First"]+list(df_csv.columns)[1:]

In [ ]:
df_csv.head()

First         B   C   D       E     F           G          H  \
0  03/10/2016 16:00:00  0.662052  16  58  banana  AYPI  Rossan Pt.  54.700000   
1  03/10/2016 16:01:00  0.123787  43  57   apple  RGVS  Mullet Pen  54.216667   
2  03/10/2016 16:02:00  0.032207  62  66  banana  JOBI        Erne  54.500000   
3  03/10/2016 16:03:00  0.865759  78  60  celery  MBPL     Kilrush  52.633333   
4  03/10/2016 16:04:00  0.321649  78  52  banana  MFRO       Louth  53.083333   

           I  
0  -8.783333  
1 -10.033333  
2  -8.266667  
3  -9.483333  
4  -6.533333

## read_excel

I'm using a random excel formatted dataset from an online source - http://www.principlesofeconometrics.com/poe5/data/excel/olympics.xlsx to show how the `read_excel` function can take a url as the path to the data file.  

Metadata about the olympics dataset - http://www.principlesofeconometrics.com/poe5/data/def/olympics.def    

In [ ]:
#df_excel = pd.read_excel("http://www.principlesofeconometrics.com/poe5/data/excel/olympics.xlsx")
df_excel = pd.read_excel("https://github.com/suzannelittle/ca682i/raw/master/data/olympics-sample-data.xlsx")

In [ ]:
df_excel.head()

conum  year co_code    gdp     pop  gold  silver  bronze  medaltot  host  \
0      1    80     AFG  26.60  16.000   0.0     0.0     0.0       0.0     0   
1      1    84     AFG  29.90  17.600   NaN     NaN     NaN       NaN     0   
2      2    80     ALB   2.45   2.671   NaN     NaN     NaN       NaN     0   
3      2    84     ALB   2.66   2.897   NaN     NaN     NaN       NaN     0   
4      2    88     ALB   2.80   3.138   NaN     NaN     NaN       NaN     0   

   planned  soviet  missing  
0        0       0        0  
1        0       0        1  
2        1       0        1  
3        1       0        1  
4        1       0        1

In [ ]:
df_excel.shape

(1610, 13)

## View and Explore your data

In [ ]:
df_excel["co_code"].unique()

In [ ]:
df_excel["co_code"].unique().size

In [ ]:
AUS = df_excel[df_excel.co_code == "AUS"].sort_values("year")

In [ ]:
AUS

In [ ]:
df_excel.plot(y="gold", x="gdp", kind="scatter")

In [ ]:
df_excel.describe()

In [ ]:
df_excel.info()